File: heap/08_reorganize_string.py
"""
PROBLEM: Reorganize String (LeetCode 767)
DIFFICULTY: Medium
PATTERN: Heap - Max-Heap by character frequency

DESCRIPTION:
Rearrange characters of s so no two adjacent characters are the same.
If impossible, return empty string "".

EXAMPLE:
Input: s = "aab" -> Output: "aba"
Input: s = "aaab" -> Output: ""
Input: s = "vvvlo" -> Output: "vlvov" (one valid)

TIME COMPLEXITY: O(n log 26) = O(n)
SPACE COMPLEXITY: O(1)
"""

import heapq
from collections import Counter


class Solution:
    def reorganizeString(self, s: str) -> str:
        freq = Counter(s)
        # Max-heap: negate frequency
        max_heap = [(-cnt, ch) for ch, cnt in freq.items()]
        heapq.heapify(max_heap)

        result = []
        while len(max_heap) >= 2:
            cnt1, ch1 = heapq.heappop(max_heap)
            cnt2, ch2 = heapq.heappop(max_heap)
            result.append(ch1)
            result.append(ch2)
            if cnt1 + 1 < 0:
                heapq.heappush(max_heap, (cnt1 + 1, ch1))
            if cnt2 + 1 < 0:
                heapq.heappush(max_heap, (cnt2 + 1, ch2))

        if max_heap:
            cnt, ch = max_heap[0]
            if -cnt > 1:
                return ""
            result.append(ch)

        return "".join(result)


if __name__ == "__main__":
    sol = Solution()
    test_cases = ["aab", "aaab", "vvvlo", "a", "aabb"]
    for s in test_cases:
        print(f"s={s!r} -> {sol.reorganizeString(s)!r}")

File: heap/08_reorganize_string.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Problem Description:
Given a string s, rearrange characters so no two adjacent characters are same.
If impossible, return empty string.

Sample Input/Output:
1) s = "aab" -> "aba"
2) s = "aaab" -> ""
3) s = "vvvlo" -> one valid output: "vlvov"

Code Explanation:
Use max-heap of {frequency, character}. Pop top two most frequent chars,
append both, reduce frequency, and push back if still available.
*/

string reorganizeString(string s) {
    vector<int> freq(26, 0);
    for (char c : s) freq[c - 'a']++;

    priority_queue<pair<int, char>> maxHeap;
    for (int i = 0; i < 26; i++) if (freq[i] > 0) maxHeap.push({freq[i], char('a' + i)});

    string result;
    while (maxHeap.size() >= 2) {
        pair<int, char> first = maxHeap.top(); maxHeap.pop();
        pair<int, char> second = maxHeap.top(); maxHeap.pop();

        int f1 = first.first;
        char c1 = first.second;
        int f2 = second.first;
        char c2 = second.second;

        result.push_back(c1);
        result.push_back(c2);

        if (--f1 > 0) maxHeap.push({f1, c1});
        if (--f2 > 0) maxHeap.push({f2, c2});
    }

    if (!maxHeap.empty()) {
        pair<int, char> last = maxHeap.top();
        int f = last.first;
        char c = last.second;
        if (f > 1) return "";
        result.push_back(c);
    }
    return result;
}

int main() {
    cout << "Output 1: " << reorganizeString("aab") << "\n";
    cout << "Output 2: " << reorganizeString("aaab") << "\n";
    cout << "Output 3: " << reorganizeString("vvvlo") << "\n";
    return 0;
}
